rules.txt

```diff
@@ -0,0 +1,110 @@
+rules_version = '2';
+service cloud.firestore {
+  match /databases/{database}/documents {
+    function signedIn() {
+      return request.auth.uid != null && request.auth.uid != null;
+    }
+    
+    function isSharedWithMe() {
+      return request.auth.token.email in resource.data.sharedWith;
+    }
+    
+    function havePendingInvite() {
+    	return request.auth.token.email in resource.data.pendingInvitations;
+    }
+    
+    function updatedDocHasCorrectOwner() {
+      return request.auth.token.email == request.resource.data.owner;
+    }
+    
+    function isDocAdmin() {
+    	return request.auth.token.email in resource.data.admins;
+    }
+    
+    function updatedDocHasSameOwner() {
+      return resource.data.owner == request.resource.data.owner;
+    }
+    
+    function updatedDocHasSameTimestamp() {
+      return resource.data.created == request.resource.data.created;
+    }
+    
+    function updatedDocHasSameId() {
+      return resource.data.id == request.resource.data.id;
+    }
+    
+    function updatedDocHasSameSharing() {
+      return resource.data.sharedWith == request.resource.data.sharedWith;
+    }
+    
+    function updatedDocHasSameAdmins() {
+      return resource.data.admins == request.resource.data.admins;
+    }
+    
+    function updatedDocHasSamePendingInvites() {
+      return resource.data.pendingInvitations == request.resource.data.pendingInvitations;
+    }
+    
+    function removedSelfFromPendingInvites() {
+    	return resource.data.pendingInvitations.removeAll(request.resource.data.pendingInvitations) == [request.auth.token.email];
+    }
+    
+    function addedSelfToSharedWith() {
+    	return request.resource.data.sharedWith.removeAll(resource.data.sharedWith) == [request.auth.token.email]
+    }
+    
+    function updatedDocHasSameName() {
+    	return resource.data.listName == request.resource.data.listName;
+    }
+    
+    function updatedDocHasSameColor() {
+    	return resource.data.colorTheme == request.resource.data.colorTheme;
+    }
+    
+    function nameAndColorUnchanged() {
+    	return updatedDocHasSameName() && updatedDocHasSameColor()
+    }
+    
+    function okChangesForPending() {
+    	return 	updatedDocHasSameAdmins() &&
+    	(updatedDocHasSameSharing() || addedSelfToSharedWith()) &&
+    	(updatedDocHasSamePendingInvites() || removedSelfFromPendingInvites());
+    }
+    
+    function essentialDataIsUnchanged() {
+    	return 	updatedDocHasSameOwner() && updatedDocHasSameTimestamp() && 
+      				updatedDocHasSameId();
+    }
+    
+    function noPermissionsChanged() {
+    	return 	updatedDocHasSameAdmins() && updatedDocHasSamePendingInvites() &&
+      				updatedDocHasSameSharing();
+    }
+    
+    match /lists-lab5/{list} {
+        allow read: if signedIn() && (isSharedWithMe() || havePendingInvite());
```

Comment:
You are checking to see whether they have a pending invite, but not whether the user's email address is validated. This could cause a security hold. A shared with email address B.  Bogus user comes along and signs up with email address B and can now see/edit the shared list.